easybuild/framework/easyconfig/easyconfig.py

```diff
@@ -195,44 +217,54 @@ def get_toolchain_hierarchy(parent_toolchain):
                 cands.append({'name': depdep['name'], 'version': depdep['version'] + depdep['versionsuffix']})
                 cands.append(depdep['toolchain'])
 
-        # only retain candidates that match subtoolchain name
-        cands = [c for c in cands if c['name'] == subtoolchain_name]
-
-        uniq_subtc_versions = set([subtc['version'] for subtc in cands])
-
-        if len(uniq_subtc_versions) == 1:
-            subtoolchain_version = cands[0]['version']
-
-        elif len(uniq_subtc_versions) == 0:
-            # only retain GCCcore as subtoolchain if version was found
-            if subtoolchain_name == GCCcore.NAME:
-                _log.info("No version found for %s; assuming legacy toolchain and skipping it as subtoolchain.",
-                          subtoolchain_name)
-                subtoolchain_name = GCCcore.SUBTOOLCHAIN
-                subtoolchain_version = ''
-            # dummy toolchain: end of the line
-            elif subtoolchain_name == DUMMY_TOOLCHAIN_NAME:
-                subtoolchain_version = ''
-            else:
-                raise EasyBuildError("No version found for subtoolchain %s in dependencies of %s",
-                                     subtoolchain_name, current_tc_name)
-        else:
-            if subtoolchain_name == DUMMY_TOOLCHAIN_NAME:
-                # Don't care about multiple versions of dummy
-                _log.info("Ignoring multiple versions of %s in toolchain hierarchy", DUMMY_TOOLCHAIN_NAME)
-                subtoolchain_version = ''
+        # only retain candidates that match subtoolchain names
+        cands = [c for c in cands if c['name'] in subtoolchain_names]
+
+        for subtoolchain_name in subtoolchain_names:
```

Comment:
@bartoldeman Shouldn't we make this whole `for` loop a separate function, named something like `det_subtoolchain_version`, which we can test separately?
It could help a lot in making this code easier to understand I think...

Also, shouldn't we `break` out of this `for` loop every time we have found a subtoolchain version?

Reply:
No, don't `break`! It's a loop over the names. Once the name has been found it is added to the toolchain_hierarchy list. The version check just needs to see if there is one version only, and else raise an error. I am moving it to a separate function now and will push that.

Reply:
@bartoldeman OK, thanks for clarifying :)

Maybe mention that as a comment above the `for`:

```python
# make sure exactly one version if found for each subtoolchain